### PR TITLE
PE-635: Safe ByteCount

### PR DESCRIPTION
**Pull request by arielmelendez**
I noticed that I was using string templates to get the correct behaviors out of a few types in JSON contexts although I had added toJSON implementations for them, so I removed the template usage in those cases. Also added tests for their toJSON methods.

UNIT TESTS SAVED ME MANY TIMES IN THIS PR.

UNIT TESTS SAVE LIVES!

### CODE REVIEW

File: src/types/byte_count.test.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+			expect(bc3.equals(bc1), `${bc3} and ${bc1}`).to.be.false;
+			expect(bc2.equals(bc3), `${bc2} and ${bc3}`).to.be.false;
+			expect(bc3.equals(bc2), `${bc3} and ${bc2}`).to.be.false;
+		});
+	});
+
+	describe('toJSON function', () => {
+		it('returns the correct JSON value', () => {
+			const byteCount = new ByteCount(12345);
+			expect(byteCount.toJSON()).to.equal(12345);
```
Comment (matibat) [inline src/types/byte_count.test.ts:62]: Shouldn't it be testing that function in conjunction with `JSON.stringify`?
```suggestion
			expect(JSON.stringify({ byteCount })).to.equal('{"byteCount":12345}');
```

File: src/types/fee_multiple.ts [hunk 2]
```diff
@@ -23,8 +23,8 @@ export class FeeMultiple {
 		return this.feeMultiple;
 	}
 
-	toJSON(): string {
-		return this.toString();
+	toJSON(): number {
+		return this.feeMultiple;
```
Comment (matibat) [inline src/types/fee_multiple.ts:27]: Nice catch!

**Review by matibat — APPROVED** [review]
I like how the redundant checks in the tests & code went away

**Review by fedellen — APPROVED** [review]

File: src/types/byte_count.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
+import { Equatable } from './equatable';
+
+export class ByteCount implements Equatable<ByteCount> {
+	constructor(private readonly byteCount: number) {
+		if (!Number.isFinite(this.byteCount) || !Number.isInteger(this.byteCount) || this.byteCount < 0) {
+			throw new Error('Byte count must be a non-negative integer value!');
+		}
+	}
+
+	[Symbol.toPrimitive](hint?: string): number | string {
```
[the remark this answers is not in the document]
Comment (arielmelendez) [reply src/types/byte_count.ts:10]: I couldn't find any Typescript standards for this. I worry that clamping down the potential values could be brittle over the long term. Do you have any docs indicating that the hint is non-nullable?

[the remark this answers is not in the document]
Comment (matibat) [reply src/types/byte_count.ts:10]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/toPrimitive
It doesn't say that explicitly, but from the description:
> The function **is called with a string argument hint**, which specifies the preferred type of the result primitive value.  **The hint argument can be one of "number", "string", and "default"**.

_"...is called with a string argument hint..."_ and `string` !== `null`

File: src/types/byte_count.test.ts (hunk 1 above)
Comment (matibat) [reply src/types/byte_count.test.ts]: Cool!
I unresolved the conversation as we might want this test case for all of the newly implemented types

File: src/types/byte_count.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (matibat) [reply src/types/byte_count.ts:10]: I also tested it in the node CLI :nerd_face:

**Pull request merged**